File: array/MonotonicQueue/python/MonotonicQueue.py
from collections import deque
from typing import Sequence


def sliding_window_maximum(values: Sequence[int], window_size: int) -> list[int]:
    candidate_indices: deque[int] = deque()
    maxima: list[int] = []

    for index, value in enumerate(values):
        while candidate_indices and candidate_indices[0] <= index - window_size:
            candidate_indices.popleft()

        while candidate_indices and values[candidate_indices[-1]] <= value:
            candidate_indices.pop()

        candidate_indices.append(index)

        if index + 1 >= window_size:
            maxima.append(values[candidate_indices[0]])

    return maxima

File: array/MonotonicQueue/rust/src/monotonic_queue.rs
use std::collections::VecDeque;

pub fn sliding_window_maximum(values: &[i64], window_size: usize) -> Vec<i64> {
    let mut candidate_indices: VecDeque<usize> = VecDeque::new();
    let mut maxima: Vec<i64> = Vec::with_capacity(values.len().saturating_sub(window_size) + 1);

    for (index, &value) in values.iter().enumerate() {
        while let Some(&front) = candidate_indices.front() {
            if front + window_size <= index {
                candidate_indices.pop_front();
            } else {
                break;
            }
        }

        while let Some(&back) = candidate_indices.back() {
            if values[back] <= value {
                candidate_indices.pop_back();
            } else {
                break;
            }
        }

        candidate_indices.push_back(index);

        if index + 1 >= window_size {
            maxima.push(values[*candidate_indices.front().unwrap()]);
        }
    }

    maxima
}
